### ISSUE TITLE
Memleak in cBugId cleanup?

### ISSUE THREAD

**Issue by dwendt** [SkyLined/cBugId#9]
Hey,

Using cBugId to hit an AV target, my code is something like this:

```python
def finishedCbk(oBugId, oBugReport):
    write2file(oBugReport.sReportHTML)

while True:
    # special process setup / verifying it's running correctly
    oBugId = cBugId(
        sCdbISA = "x64",
        auApplicationProcessIds = [pid],
        fFinishedCallback = finishedCbk)

    oBugId.fStart();

    # start a fuzzer thread asynchronously here, doesn't interact with oBugId

    oBugId.fWait();

    # ensure teardown of target process

    # tried calling fStop here, del oBugId, etc. no difference

    # dump largest str objects in memory, compare to strs from last time.
```

On that last line I'm seeing the largest string as always around 5mb, and it appears to be a portion of the htmlreport related to the CDB command listing. It's the same kind of snippet across each time this loops. It never seems to get garbage collected, even when calling gc.collect().  A dump of the first few thousand chars is below:

```html
<hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">.prompt_allow -dis -ea -reg -src -sym;</span> <span class="CDBComment">$ Display only the prompt</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">.pcmd -s &quot;.echo;&quot;;</span> <span class="CDBComment">$ Output a CRLF after running the application</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">.lastevent;</span> <span class="CDBComment">$ Get information about last event</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">!peb;</span> <span class="CDBComment">$ Get current proces environment block</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">lmov a 0x7FF62C250000;</span> <span class="CDBComment">$ Get module information</span><br/><br/><hr/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">.childdbg 1;</span> <span class="CDBComment">$ Debug child processes</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">sxd *;sxi ld;sxi ud;sxd 0xC0000094;sxd 0xC0000095;sxd 0xC0000008;sxd 0xC0000235;sxd 0x80000004;sxd 0x4000001E;sxd 0x40080201;sxd 0xE06D7363;sxe cpr;sxe ibp;sxe epr;sxe aph;sxe 0xC0000005;sxe 0xC0000420;sxe 0x80000003;sxe 0xC000008C;sxe 0x80000002;sxe 0xC0000602;sxe 0x80000001;sxe 0xC000001D;sxe 0xC0000006;sxe 0xC0000096;sxe 0xC0000409;sxe 0xC00000FD;sxe 0x4000001F;sxe 0x80000007;</span> <span class="CDBComment">$ Setup exception handling</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">~*m;</span> <span class="CDBComment">$ Resume all threads</span><br/><br/><hr/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">!heap -p;</span> <span class="CDBComment">$ Get page heap status</span><br/><br/><hr/><span class="CDBPrompt">0:018&gt; </span><span class="CDBCommand">.time;</span> <span class="CDBComment">$ Get debugger time</sp
```

I'm sure you can spot where the GC lifetime of whatever that is should be, if not I can do more work on figuring out python to track down its refs.

**Comment by SkyLined** [comment]
Thanks for the report. I never looked into memory usage, so I'm sure there are a few things that could be improved. Unfortunately, this also means I have no idea what is causing this. What you are looking at is indeed the log of commands sent to cdb that is stored in the HTML report. It is stored in the `sCdbIOHTML` property of the `cCdbWrapper` instance by the `cCdbWrapper_fasExecuteCdbCommand` and `cCdbWrapper_fasReadOutput` functions.

cBugId is multi-threaded, which means there could be dangling references in a different thread; there are three "vital" threads that run while cdb.exe is running: one to process stdin/stdout, one to process stderr and one that waits until both have died before doing some clean up. They are started when you call `cBugId.fStart`. However, these three threads must terminate for cBugId to stop: after oBugId.fWait() none of these should be left running. This means these threads should not be able to hold a reference at the point where you are seeing the memory leak, so I do not believe this is the cause.

There are also helper threads: I create one every time the application is running in order to check if a timeout must be fired and the application interrupted. This is done in `cCdbWrapper_fasReadOutput`. However, that functions waits for this thread to terminate before returning (in a finally clause), so I don't think that can be the cause either.

So, as far as I know, there is only one thread running after `oBugId.fWait()` returns and the only references to the data you see would be through `oBugId.__oCdbWrapper.sCdbIOHTML`. If you delete oBugId, the string should be gone. Note that that string is also part of the HTML report (which is one large string), which would be referenced through `oBugId.oBugReport.sReportHTML`, which should also be gone if you delete oBugId.

In other words: I have no idea how this can happen. If you could track down the refs that keep it from being garbage collected, that would really help!

Thanks

**Comment by SkyLined** [comment]
I am going to close this issue as I don't think it's a big enough problem for anyone to invest time in analyzing and/or fixing it.
